### DISCUSSION TITLE
Support for Redis script for multiple shard

### DISCUSSION THREAD

**Discussion by aratnamirasugab** [redisson/redisson#6037] [Q&A] [answered]
Hi, 

wanted to ask if there's any way to execute Lua script that will fetch all keys from multiple shard.  

Background story : 
I'd like to scan all keys on 5 shard by using eval script. my goal is to update all keys if they don't have TTL and re-set the TTL. 

My solution was to create Lua script, check the existing TTL, and if the TTL is -1, we will update the TTL.

First operation
```
{
    String luaScript =
        "local result = {} " +
        "local cursor = ARGV[1] " +
        "local batchSize = ARGV[2] " +
        "repeat " +
        "local scanResult = redis.call('SCAN', cursor, 'MATCH', '*', 'COUNT', batchSize) " +
        "cursor = scanResult[1] " +
            "for i, key in ipairs(scanResult[2]) do " +
            "if #result < tonumber(batchSize) then " +
            "table.insert(result, key) " +
            "end " +
        "end " +
        "until #result >= tonumber(batchSize) " +
        "return {cursor, result}";

    return redissonClient
        .getScript(StringCodec.INSTANCE)
        .eval(
            RScript.Mode.READ_ONLY,
            luaScript,
            RScript.ReturnType.MULTI,
            List.of(), cursor, batchSize
        );
  }
```

Second operation
```
{
    return redissonClient.getKeys().remainTimeToLive(key);
  }
```

Third operation
```
{
    RBucket bucket = redissonClient.getBucket(key);
    return bucket.expire(ttlMilliseconds, TimeUnit.MILLISECONDS);
  }
```

But I think I missed something, the whole operation was done for 2 hour and after I checked the metrics on the Redis Cluster, there was only 1 Redis shard that updated. I am sure that since only the 1st shard that increased on the metrics.

metrics that went up during the operation : 
- Network Packets In
- Network Packets Out
- DB0 Average TTL
- Cache Hit Rate

```
Redis version : 7.1.0
Cluster mode : Enabled
Shards : 5
redisConfig.useClusterMode : true
Redis Client config using the cluster endpoint

```

Is there any way to do the operation for the other shards? 

Thanks!

**Comment by mrniko** [comment]
Unfortunately, there is no such option. 

I'm considering adding `eval(Mode mode, String luaScript, ReturnType returnType, Object... values)` method. It will be executed on all master or slave nodes in Redis cluster.

Here is the ticket https://github.com/redisson/redisson/issues/5946

**Reply by aratnamirasugab** [reply]
Hi @mrniko thanks for replying.

I see, so in the meantime, I guess the only solution is fetch the keys from each shard/node directly using its endpoint then. 

Or perhaps if you have anything to share please let me know. 🙇 

Thanks!

**Reply by mrniko** [reply]
Why can't you use org.redisson.api.RKeys#getKeys(int) method instead of the first script?

**Reply by mrniko** [reply, marked as answer]
The ticket https://github.com/redisson/redisson/issues/5946 has been resolved.
